[libclang] Add API to generate a reproducer for the explicitly-built modules.

vsapsai opened the pull request:
Capturing a single `-cc1` command to reproduce a bug with the explicitly-built modules isn't sufficient, we need to know what .pcm files were built and how.

rdar://59743925

vsapsai committed:
[libclang] Add API to generate a reproducer for the explicitly-built modules. (#10577)

Capturing a single `-cc1` command to reproduce a bug with the explicitly-built
modules isn't sufficient, we need to know what .pcm files were built and how.

rdar://59743925

vsapsai commented:
Right now I'd like to get a high-level opinion if such an approach for reproducers is reasonable. This feature will have more commits, to get an idea of how it is supposed to look like you can check [users/vsapsai/reproducer-explicitly-built-modules](https://github.com/swiftlang/llvm-project/commits/users/vsapsai/reproducer-explicitly-built-modules/) branch.

vsapsai commented:
An example of API usage is in https://github.com/swiftlang/swift-build/pull/455

vsapsai commented:
By the way, this doesn't work as a reproducer for dependency scanning, I'm just capturing the scanning results. My hope that with including the original command [used for scanning] and with the captured files you can still do something like
```
c-index-test core --scan-deps -- original_command_from_reproducer
```

After landing the proposed level of reproducing we can think how to support scan-deps better.

I was considering an alternative of invoking clang-scan-deps from the shell script and building modules based on the scanning results. But I've decided that embedding a little build system in a shell script is too much complexity and not justified at this stage. Maybe we'll see a need for something like that, then we can bite the bullet and add the required complexity.

vsapsai commented:
I had an idea of adding `DependencyScanningWorker` to the reproducer generation API so that we can capture some internal worker state to help with the debugging. But after additional consideration I think it is too vague and unclear as I don't have specific use cases when it should help. And such functionality we can always add in a new API version.

vsapsai commented:
Ping.

jansvoboda11 commented:
I'd like the API changes from [users/vsapsai/reproducer-explicitly-built-modules](https://github.com/swiftlang/llvm-project/compare/users/vsapsai/reproducer-explicitly-built-modules) (and the addition of `const char *ModuleName` to align with the scan API that we talked about offline) to be included in this PR. C APIs cannot change between Clang versions, so getting this right in the first commit makes the most sense to me.

vsapsai committed:
[libclang] Allow to put a reproducer into a custom location. (#10577)

Tools using the reproducers can specify a custom location to support
scenarios similar to those covered by `-fcrash-diagnostics-dir` and
`CLANG_CRASH_DIAGNOSTICS_DIR`.

vsapsai commented:
Added `ModuleName` [that's not used right now]. And added another commit with `ReproducerLocation` to this branch.

jansvoboda11 reviewed (approved):
LGTM, but let's see if @Bigcheese or @benlangmuir have any objections.

benlangmuir commented on clang/include/clang-c/Dependencies.h:
```diff
@@ -325,6 +325,38 @@ CINDEX_LINKAGE void
+ * for a success case but is allowed to be empty when encountered an error.
+ */
+CINDEX_LINKAGE enum CXErrorCode
+clang_experimental_DependencyScanner_generateReproducer(
```
I wonder if this should use `CXDependencyScannerWorkerScanSettings` instead of `int argc, const char *const *argv, const char *ModuleName,
     const char *WorkingDirectory` for the basic compilation setup.

How sure are we that `const char *ReproducerLocation,
     bool UseUniqueReproducerName,` are the only two reproducer-specific options we will ever want here?

vsapsai replied:
> I wonder if this should use `CXDependencyScannerWorkerScanSettings` instead of `int argc, const char *const *argv, const char *ModuleName, const char *WorkingDirectory` for the basic compilation setup.

I was going back and forth about it and can be persuaded either way. My [weak] justification is that we don't pass the settings to a dependency service/worker/tool. So it's not a self-sufficient entity but more a convenience container that happens to be the same between different functions. So I avoid treating it as an independent entity.

> How sure are we that `const char *ReproducerLocation, bool UseUniqueReproducerName,` are the only two reproducer-specific options we will ever want here?

For the potential enhancements I expect only something to help capture the scanner state. Unfortunately, I don't know what that would be. Maybe .pcm cache, maybe dumping stat cache, maybe something else. So far it is too vague and not the most important direction for the next steps. That's why I'd like to land something reasonably future-proof and iterate on it to see what else we might need.

benlangmuir replied:
> My [weak] justification is that we don't pass the settings to a dependency service/worker/tool.

Isn't that exactly what we do? The difference is that you're creating a new one instead of using an existing one.

> hat's why I'd like to land something reasonably future-proof and iterate on it to see what else we might need.

If you want to make this future proof the C API should probably be much more opaque about the configuration, similar to what we do with the worker scan settings where the settings become their own opaque type and we can add new C functions over time to add new settings to that configuration object without breaking the existing API.

vsapsai replied:
> > My [weak] justification is that we don't pass the settings to a dependency service/worker/tool.
> 
> Isn't that exactly what we do? The difference is that you're creating a new one instead of using an existing one.

I mean that in `clang_experimental_DependencyScannerWorker_getDepGraph`, for instance, `Settings.MLO` and `Settings.MLOContext` goes into `DependencyScanningTool::createActionController`. `Settings.argc`, `Settings.argv`, `Settings.WorkingDirectory` goes into `computeDependencies`. And `Settings.ModuleName` affects the function's logic. We don't use `Settings` as a single entity anywhere and taking it apart is the first thing we do.

> > hat's why I'd like to land something reasonably future-proof and iterate on it to see what else we might need.
> 
> If you want to make this future proof the C API should probably be much more opaque about the configuration, similar to what we do with the worker scan settings where the settings become their own opaque type and we can add new C functions over time to add new settings to that configuration object without breaking the existing API.

Do you suggest using `CXDependencyScannerWorkerScanSettings` or introducing something like `CXDependencyScannerReproducerSettings`? I believe a single function with simple arguments is easier to work with and introducing `clang_experimental_DependencyScanner_generateReproducer_v2` when necessary is not a big deal. But it's not a belief worth fighting, so I'm fine using whatever settings wrapper you think would be better.

benlangmuir replied:
> Settings.argc, Settings.argv, Settings.WorkingDirectory goes into computeDependencies. And Settings.ModuleName affects the function's logic.

The use of `getTranslationUnitDependencies` is basically the same here as the use of `computeDependencies` elsewhere.  The fact you're not using the lookup callback could be a possible reason not to use `CXDependencyScannerWorkerScanSettings`.

Incidentally, why do you accept `ModuleName` and not use it?

> Do you suggest using CXDependencyScannerWorkerScanSettings or introducing something like CXDependencyScannerReproducerSettings? I believe a single function with simple arguments is easier to work with and introducing clang_experimental_DependencyScanner_generateReproducer_v2 when necessary is not a big deal. But it's not a belief worth fighting, so I'm fine using whatever settings wrapper you think would be better.

I'm trying to understand what kind of changes are likely to come to this API over time, how it fits with the other scanner APIs, and so what makes sense to make it stable.  Introducing a v2 is not necessarily a big deal, but one of these APIs made it to v5 very quickly and was annoying to work with in client code that needs to support multiple versions of the dylib, like swift-build does.  Having explicit opaque settings objects is harder to work with when you have one version of an API, but it's easier to factor if we are supporting N versions of the API. 

If you are confident that this API will not need more parameters, then I'm fine with keeping it like it is and we can make it opaque in a hypothetical v2 if there is an unexpected need to add one.  If we already expect there is a good chance of needing changes to the API in future, then I'm inclined to make it opaque from the start.

vsapsai replied:
> Incidentally, why do you accept `ModuleName` and not use it?

To be honest, I've added it mostly because Jan asked about it. I can see there might be a noticeable enough divergence between for-TU scanning and for-module scanning in the future, so it can be useful. But I'm not using it now because there is no data indicating it is useful. So this is "let's try and see how it works" approach. Having `ModuleName` in API seems like a plausible enough future-proofing step.

> I'm trying to understand what kind of changes are likely to come to this API over time, how it fits with the other scanner APIs, and so what makes sense to make it stable. Introducing a v2 is not necessarily a big deal, but one of these APIs made it to v5 very quickly and was annoying to work with in client code that needs to support multiple versions of the dylib, like swift-build does. Having explicit opaque settings objects is harder to work with when you have one version of an API, but it's easier to factor if we are supporting N versions of the API.
> 
> If you are confident that this API will not need more parameters, then I'm fine with keeping it like it is and we can make it opaque in a hypothetical v2 if there is an unexpected need to add one. If we already expect there is a good chance of needing changes to the API in future, then I'm inclined to make it opaque from the start.

Fair enough, that v5 doesn't give off good vibes.

I think the main question to you would be if you see a need for some CAS special handling?

As for the known problems (that we need to be prepared to fix) we have
* Failing reproducer commands, e.g., unescaped file paths with spaces. No API change required.
* No integration with Swift. No API change required based on how scanning is integrated with Swift.
* Non-relocatable precompiled headers. No API change required.
* Unknown worker state that causes crashes/errors. Might require API change.
* Stale built modules. Might require API change.

The last two are a little bit of a stretch because reproducers don't promise to help with reproducing intermittent issues. If you cannot reproduce an issue in a clean build, there is no magic in reproducers that makes intermittent issues predictable.

It is possible there are unknown issues that would require API changes but as far as I know, API is fairly stable in its current shape.

benlangmuir replied:
> I think the main question to you would be if you see a need for some CAS special handling?

Good question. If you want reproducers to work for cached compilations, we need a CAS and CAS configuration. It's not obvious to me how we want to deal with that.
* You could create a new on-disk CAS for the reproducer.
  * Advantages: it would only contain what is needed for the reproducer (much smaller) and would be standalone. 
  * Disadvantages: it will not work if you import a PCH that was produced with a different existing CAS; if somehow the crash to be reproduced depends on the specific CAS configuration, it will not be there.
* You could take `CXCASOptions` and `CXCASDatabases` parameters, similar to how `CXDependencyScannerServiceOptions` is set up.
  * Advantages: you can get whatever CAS you want
  * Disadvantages: the client (build system?) needs to know how to configure it for the reproducer. The reproducer may not be standalone if it depends on an external CAS.

vsapsai replied:
The general approach right now is that reproducers aren't to help debug issues in any specific area but just to allow to reproduce and to share a compilation. That's why I don't plan to include a dependency worker in the API, even if it can be useful to debug issues with the dependency scanning. From this perspective I should go with the first approach "create a new on-disk CAS for the reproducer".

But given that we might need to add some CAS options in the future, now I'm leaning heavily towards adding `CXDependencyScannerReproducerSettings`. We might not need it but we have 2 use cases for it, even if not fully fleshed out.

If there is nothing required in API to prepare for "create a new on-disk CAS for the reproducer", I'll just work on adding `CXDependencyScannerReproducerSettings`.

vsapsai replied:
Small naming adjustment - Options seems to be noticeably more popular than Settings. So will go with `CXDependencyScannerReproducerOptions`.

benlangmuir replied:
SGTM, other than the storage issue I mentioned on the new options object, this PR now LGTM.

vsapsai replied:
Done now.

benlangmuir commented on clang/tools/libclang/CDependencies.cpp:
```diff
@@ -317,6 +317,152 @@ void clang_experimental_DependencyScannerWorkerScanSettings_dispose(
+  DependencyScanningService DepsService(
+      ScanningMode::DependencyDirectivesScan, ScanningOutputFormat::Full,
+      CASOpts, /*CAS=*/nullptr, /*ActionCache=*/nullptr, FS);
+  DependencyScanningTool DepsTool(DepsService);
```
Why does this API create its own service/worker/tool instead of taking a `CXDependencyScannerWorker`?

vsapsai replied:
> Why does this API create its own service/worker/tool instead of taking a `CXDependencyScannerWorker`?

I wanted the reproducer scan to be independent from previous scans, that's why don't use an existing worker. It is a trade-off, not a uniquely correct approach. Reproducers should be useful in general, not only for debugging issues with EBM, that's why I'm leaning towards from-scratch scanning. And I hope it is easier to reason about than to take into account an invisible worker's state.

benlangmuir replied:
> independent from previous scans

Can you clarify what specifically you want to be independent from?  Is it the file system cache?

vsapsai replied:
It's less about reusing specific data but more about creating API that is easier to reason about. For example, I don't need to know implementation details that `CXDependencyScannerWorker` doesn't maintain state that affects the reproducer results. That generating a reproducer doesn't affect `CXDependencyScannerWorker` state in unpredictable way. That I don't need to be concerned about `CXDependencyScannerWorker` being accessed concurrently. In practice, passing `CXDependencyScannerWorker` works despite all of these concerns but there is no guarantee it won't change in the future.

benlangmuir commented on clang/tools/libclang/CDependencies.cpp:
```diff
@@ -317,6 +317,152 @@ void clang_experimental_DependencyScannerWorkerScanSettings_dispose(
+  llvm::SmallString<128> ReproScriptPath;
+  int ScriptFD;
+  if (ReproducerLocation) {
+    if (!llvm::sys::fs::exists(ReproducerLocation)) {
```
`create_directories` has `bool IgnoreExisting = true`

benlangmuir reviewed (commented):
Should this have an entry point in clang-scan-deps? It's unusual to have something in CDependencies that doesn't have an equivalent clang-scan-deps command-line interface, though it's not strictly necessary

vsapsai commented:
> Should this have an entry point in clang-scan-deps? It's unusual to have something in CDependencies that doesn't have an equivalent clang-scan-deps command-line interface, though it's not strictly necessary

I agree it is a good idea. But when I've started working on it, I've realized that for the proper API I need a build system ready to use it. And I don't have such a build system. I'm not strongly against it but I'd rather avoid developing functionality nobody uses.

benlangmuir commented on clang/tools/libclang/CDependencies.cpp:
```diff
@@ -640,3 +640,176 @@ std::string OutputLookup::lookupModuleOutput(const ModuleDeps &MD,
 }
+
+namespace {
+struct DependencyScannerReproducerOptions {
```
We need to copy the strings/arrays so that their lifetime is maintained by the options object.
